RFC: start rusting the ADT code

chadmed opened the pull request:
Don't really see a point to merging this until it's done. This is just to track the work and make sure I'm on the right track.

svenpeter42 commented on Makefile:
```diff
@@ -67,10 +67,9 @@ endif
 export RUSTC_BOOTSTRAP=1
 RUST_LIB := librust.a
-RUST_LIBS :=
+RUST_LIBS += $(RUST_LIB)
```
You can drop RUST_LIBS entirely by using RUST_LIB inside OBJECTS directly

chadmed commented:
Rebased on top of main, and refactored API to make a bit more sense on the Rust side.

`time proxyclient/tools/chainload.py build/m1n1.macho` 
on main: 2.171s
on riir/adt: 2.144s

WhatAmISupposedToPutHere commented on rust/src/adt.rs:
```diff
@@ -0,0 +1,846 @@
+
+const ADT_ALIGN: i32 = 4;
+
+#[repr(C)]
```
Fat pointers like &[T] do not have a stable C representation despite rust allowing you to put #[repr(C)] on a struct containing one.

chadmed replied:
Better to declare it as `*const u8` then? The struct doesn't actually do anything, it was a leftover from a previous architecture. Probably fine to just get rid of it and move the functions in the ADT impl out of it.

WhatAmISupposedToPutHere commented on rust/src/adt.rs:
```diff
@@ -0,0 +1,846 @@
+const ADT_ALIGN: i32 = 4;
+
+#[repr(C)]
+pub struct ADT<'a> {
```
Is it even constructed anywhere? looks like a dead struct.

WhatAmISupposedToPutHere commented on rust/src/adt.rs:
```diff
@@ -0,0 +1,846 @@
+    }
+
+    /// Trace a given ADT path and return the offset of the final node.
+    pub fn path_offset_trace(path: &str, mut offsets: Option<&mut [i32]>) -> Result<i32, AdtError> {
```
Here and everywhere else where you use node offsets - unsound api from Rust's POV as the offsets are invalidated when you change the ADT, but i am not sure how to properly model it for now and whether it is worth it.

chadmed commented:
I've changed the internal API to use absolute pointers to the nodes/properties instead of offsets. The C bindings calculate these relative to the head of the ADT, then derives offsets to give back to C code.

The public Rust-side API deals with neither pointers nor offsets. Instead, tracing a path will yield a slice of static references where required, which can be indexed into/iterated over/whatever. Internally we still need to deal with raw pointers since all we get from iBoot is a pointer to the head of the ADT, however this can be hidden from consumers now.

WhatAmISupposedToPutHere commented on rust/src/adt.rs:
```diff
@@ -0,0 +1,909 @@
+/// Determine if an ADT node's name is equal to some string bounded by a length.
+/// This is required to match the semantics expected when doing a full ADT path
+/// trace.
+fn node_names_equal(a: &str, b: &str, len: usize) -> bool {
```
pre-slice b in the caller?

chadmed replied:
Got rid of the need to slice the strings at all

WhatAmISupposedToPutHere commented on rust/src/adt.rs:
```diff
@@ -0,0 +1,909 @@
+) -> Result<(u64, u64), AdtError> {
+    let head = unsafe { ADTNode::from_ptr(adt as *const ADTNode)? };
+
+    if nodes[0].is_some() {
+        if nodes[0].unwrap().as_ptr() == unsafe { adt as *const u8 } {
```
if let Some(...) = nodes[0]

chadmed replied:
Added sanity check made it safe to just unwrap nodes[0] unconditionally

WhatAmISupposedToPutHere commented on rust/src/adt.rs:
```diff
@@ -0,0 +1,909 @@
+    }
+    let mut cursor: usize = nodes.len() - 1;
+
+    while nodes[cursor].is_none() {
```
panics if nodes does not contain at least one Some. Can that happen?

chadmed replied:
Added a sanity check to make sure we have at least one Some

WhatAmISupposedToPutHere commented on rust/src/adt.rs:
```diff
@@ -0,0 +1,909 @@
+        let p_cnt = self.property_count;
+        let mut p = self.first_property()?;
+
+        for _i in 0..p_cnt {
```
`s/_i/_/` 

same below

WhatAmISupposedToPutHere commented on rust/src/adt.rs:
```diff
@@ -0,0 +1,909 @@
+        let mut cnt = self.child_count;
+        let mut c = self.first_child()?;
+
+        while cnt > 0 {
```
for instead of while?

WhatAmISupposedToPutHere reviewed (commented):
lgtm in general, mostly some minor style nitpicks
